crates/matrix-sdk-crypto/src/session_manager/group_sessions/share_strategy.rs

```diff
@@ -225,6 +241,41 @@ fn split_recipients_withhelds_for_user(
     RecipientDevices { allowed_devices: recipients, denied_devices_with_code: withheld_recipients }
 }
 
+fn split_recipients_withhelds_for_user_based_on_identity(
+    user_devices: HashMap<OwnedDeviceId, ReadOnlyDevice>,
+    device_owner_identity: &Option<ReadOnlyUserIdentities>,
+) -> RecipientDevices {
+    let mut allowed_devices: Vec<ReadOnlyDevice> = Default::default();
+    let mut denied_devices_with_code: Vec<(ReadOnlyDevice, WithheldCode)> = Default::default();
+
+    match device_owner_identity {
+        None => {
+            // withheld all the users devices, we need to have an identity for this
+            // distribution mode
+            denied_devices_with_code
```

Comment:
Let's return in each branch of that match a `RecipientDevices` struct, makes it a bit clearer that we're not really mutating multiple times the above vectors and might get rid of a couple of allocations.

```suggestion
            RecipientDevices {
                denied_devices_with_code: user_devices
                    .into_values()
                    .map(|d| (d, WithheldCode::Unauthorised)),
                ..Default::default()
            }
```
(You won't be able to apply that suggestion because I can't modify multiple lines and you need further modifications in the rest of the function).